### ISSUE TITLE
Bundling jobs smaller than node size

### ISSUE THREAD

**Issue by csadorf** [glotzerlab/signac-flow#31]
**[Original report](https://bitbucket.org/glotzer/signac-flow/issue/31) by Michael Howard (Bitbucket: [mphoward](https://bitbucket.org/mphoward), GitHub: [mphoward](https://github.com/mphoward)).**

----------------------------------------

I've run into an issue using the `--bundle` and `--parallel` options of the submit interface for an operation in my project. After my simulations are complete, I want to run an analysis on each job that uses only 1 core. Since I need to take a whole node, I'd like to place 16 of these operations per node and execute in parallel. The submit command generates a script like I would expect, where multiple commands are submitted in background followed by a wait.

This works well with slurm + srun, but I am running on Blue Waters with torque + aprun. I noticed that my jobs were hitting the walltime, even though there should have been no problem. I dug into the Blue Waters documentation and found that, apparently, only one aprun command is permitted on a compute node at a time (https://bluewaters.ncsa.illinois.edu/job-bundling). I believe this either serializes or kills the other aprun jobs (not sure).

The solution, as suggested in the Blue Waters docs, is to either use another tool or to use xargs to run multiple commands. I opted to use xargs, since I could dump out the submission script to file and then modify it by moving lines around. This is obviously frustrating and error prone, especially as more tasks are run.

Is there any way of customizing how jobs are bundled by environment? Or, could there be a mechanism for selecting from a few "standard" bundle options for different schedulers?

**Comment by vyasr** [comment]
This seems closely related to #114, #52, and #62. The solution should be easy once those are merged in since the different operations could be easily submitted as one aprun command.

**Comment by bdice** [comment]
I think this issue can be closed. Some clusters like Stampede2 offer utilities like "Launcher", which signac-flow supports (at least in theory; I have not tested it). Handling a general case of this kind of executor-bundling is out of scope in my view, because there are so few commonalities between clusters. The FlowGroup functionality in signac-flow 0.10 and future support for aggregation (in progress) provides further workflow execution modes that also touch on some of the core user needs in the original issue. Users who need a solution similar to Stampede2's Launcher can use that code as a reference and adapt for their own cluster.

**Comment by vyasr** [comment]
I think once aggregation is merged we can treat this issue as solved. A combination of aggregation, groups, and bundling should be sufficient for users to customize submissions as needed. No need to reopen this issue though, aggregation will be resolved soon enough.
